jp.go

```diff
@@ -71,6 +78,21 @@ func runMain(c *cli.Context) int {
 		}
 		expression = c.Args()[0]
 	}
+	// NoColor defines if the output is colorized or not. It's dynamically set to
+	// false or true based on the stdout's file descriptor referring to a terminal
+	// or not. It's also set to true if the NO_COLOR environment variable is
+	// set (regardless of its value). This is a global option and affects all
+	// colors.
+	switch c.String("color") {
+	case "always":
+		color.NoColor = false
+	case "auto":
+		// this is the default in the library
+	case "never":
+		color.NoColor = true
```

Comment:
Why are you mutating `NoColor` ?

Reply:
I have no idea – I’m just trying to replicate [an existing](https://github.com/jmespath/jp/pull/13) _pull request_.

Basically, the way I understand it is:

- The underlying library is automatically setting `color.NoColor` to either `true` or `false` depending upon
  - The stdout file descriptor being a TTY
  - Or the value of the `NO_COLOR` environment variable.
- The command-line flag allows a user to honor that value, by setting it to `auto` or override this behaviour.
  - Setting the `--colors` command-line flag to `never` will force it to `true`.
  - Setting the `--colors` command-line flag to `always` will force it to `false`

I think that makes sense.

Reply:
Ok, it feels strange but maybe you're allowed to mutate it 🤷

Reply:
I think there’s no good answer as the underlying library as a behaviour that depends on a global environment variable...

Reply:
From what I understand, if `c.noColor` is set it takes precedence over the global `NoColor`

Reply:
I have tried to write code like this, but it actually produces no output at all if colors are disabled:

```go
func setColoredOutput(enabled bool) {

	if enabled {
		jsoncolor.DefaultArrayColor.EnableColor()
		jsoncolor.DefaultColonColor.EnableColor()
		jsoncolor.DefaultCommaColor.EnableColor()
		jsoncolor.DefaultFalseColor.EnableColor()
		jsoncolor.DefaultFieldColor.EnableColor()
		jsoncolor.DefaultFieldQuoteColor.EnableColor()
		jsoncolor.DefaultNullColor.EnableColor()
		jsoncolor.DefaultNumberColor.EnableColor()
		jsoncolor.DefaultObjectColor.EnableColor()
		jsoncolor.DefaultSpaceColor.EnableColor()
		jsoncolor.DefaultStringColor.EnableColor()
		jsoncolor.DefaultStringQuoteColor.EnableColor()
		jsoncolor.DefaultTrueColor.EnableColor()

	} else {
		jsoncolor.DefaultArrayColor.DisableColor()
		jsoncolor.DefaultColonColor.DisableColor()
		jsoncolor.DefaultCommaColor.DisableColor()
		jsoncolor.DefaultFalseColor.DisableColor()
		jsoncolor.DefaultFieldColor.DisableColor()
		jsoncolor.DefaultFieldQuoteColor.DisableColor()
		jsoncolor.DefaultNullColor.DisableColor()
		jsoncolor.DefaultNumberColor.DisableColor()
		jsoncolor.DefaultObjectColor.DisableColor()
		jsoncolor.DefaultSpaceColor.DisableColor()
		jsoncolor.DefaultStringColor.DisableColor()
		jsoncolor.DefaultStringQuoteColor.DisableColor()
		jsoncolor.DefaultTrueColor.DisableColor()
	}
}
```

I think the cleanest way would be to create a formatter with either default colors as set by the library, or all colors set to white.

I think the code using the global `noColor` is good for now. WDYT?

Reply:
<img width="524" alt="image" src="https://user-images.githubusercontent.com/47974576/226452538-a515cb35-c3cf-4546-ad06-3c2aaaac0ed0.png">

Reply:
@springcomp shall i push in your pr ?

Reply:
Please, let me do this properly with a formatter.

Reply:
Why do you want to use a formatter ?

Reply:
If that's ok to mutate `color.noColor` I'll just rethink the function pointer-based code with a simple if condition.

If you think mutating a global variable is not a good practice, to which I agree, I think a formatter is the way to go.

Reply:
But EnableColor/DisableColor works fine.

Reply:
I found that you have to `EnableColor()` on each individual color one at a time.
Or am I missing something?

Reply:
I will push my branch so that you can give it a try

Reply:
@springcomp see https://gist.github.com/eddycharly/ee6cceff6b4baea88aa7aa0cee04c43e

Reply:
@eddycharly Thanks !

That’s what I somehow ended up with approximately.

I had to fix a bug in the [nwidger/jsoncolor](https://github.com/nwidger/jsoncolor/issues/7) library to make tests pass.
For the time being, I’m using my own fork to workaround the issue, but I have concerns over maintainance in the future.

WDYT?

Reply:
The issue is that [nwidger/jsoncolor] will always add a space character before JSON scalar values.
So, for instance, given:

```json
{ "foo": "bar" }
```

The expression `` foo `` will return `` _"bar" `` (with a leading space) instead of `` "bar" ``.
This broke the tests.

Although this is a minor issue, I’m quite a nitpicky sort of guy so I wanted to make sure the output would be identical.

Reply:
Definitely, I would report the bug (and eventually open a PR)

Reply:
Of course, I [did that first](https://github.com/nwidger/jsoncolor/pull/8) .